### ISSUE TITLE
Is there something wrong when computing mIOU?

### ISSUE THREAD

**Issue by zhouzg** [IvLabs/stagewise-knowledge-distillation#12]
Hi, thanks for your sharing!
When computing the mIOU of the whole dataset,  it seems that you just average the miou of each batch , but I think it should not be that.

    # stagewise-knowledge-distillation\semantic_segmentation\utils\metrics.py
    def mean_iou(model, dataloader, args):
        gpu1 = args.gpu
        ious = list()
        for i, (data, target) in enumerate(dataloader):
            data, target = data.float().to(gpu1), target.long().to(gpu1)
            prediction = model(data)
            prediction = F.softmax(prediction, dim=1)
            prediction = torch.argmax(prediction, axis=1).squeeze(1)
            ious.append(iou(target, prediction, num_classes=11))
        print(len(ious))
        return (sum(ious) / len(ious))

**Comment by akshaykvnit** [comment]
We are calculating the average over the entire dataset. The loop over the dataloader keeps adding the mious of each batch to the list and we take the average at the end, which gives us the mean IoU over the entire dataset.

**Comment by zhouzg** [comment]
I think the mean IoU over the entire dataset should be calculated as follows:
1、for each class,  calculate the total area_intersection and area_union  over the entire dataset
2、for each class, iou = total area_intersection / total area_union 
3、MIOU = mean of iou for each class
In author's code, miou will be changed with different batch size.

**Comment by akshaykvnit** [comment]
Thank you for informing us about this problem. We have surveyed several repositories and websites related to this (which we list below). However, we could not find any standard implementation or guidelines for mIoU computation. However, we observed that both methods are in use.

Our method -
https://github.com/tatigabru/kaggle-lyft/blob/master/src/validation.py
https://github.com/penny4860/Kitti-road-semantic-segmentation/blob/master/eval.py (commented out, but used since they mention the score in readme)
https://github.com/vilmar-hillow/kaggle_tgs_salt/blob/master/evaluate.py
https://github.com/gestalt-howard/tgs-salt-identification/blob/master/utils/evaluate.py
https://github.com/KokeCacao/RedstoneTorch/blob/master/eval.py
https://www.kaggle.com/c/tgs-salt-identification-challenge/discussion/63044

Your suggested method - 
https://github.com/thpss92093/pytorch_fcn/blob/master/sis_fcn_products_20.ipynb
https://github.com/MarritL/GrasslandProject/blob/ffdfd5bedb4e16acb3acf86a9cd0b4105df4465b/pytorch/evaluate_testset.py
https://github.com/VainF/DeepLabV3Plus-Pytorch/blob/master/metrics/stream_metrics.py
https://github.com/chenxi116/DeepLabv3.pytorch/blob/master/main.py

As far as the results of this work are concerned, we have used our implementation consistently for all our experiments (even those implemented for comparing with other works). Thus, the results presented by us remain comparable and show the usefulness of the proposed method. Regarding the changing results with batch size, we have performed all evaluations using the same batch size and thus the results are comparable. However, yes, we agree that results will change with batch size (though we cannot comment on whether it will be a significant change or not without actually testing it).
